libraries/AP_BattMonitor/AP_BattMonitor_AD7091R5.cpp

```diff
@@ -0,0 +1,236 @@
+    // @DisplayName: Voltage Multiplier
+    // @Description: Used to convert the voltage of the voltage sensing pin (@PREFIX@VOLT_PIN) to the actual battery's voltage (pin_voltage * VOLT_MULT). 
+    // @User: Advanced
+    AP_GROUPINFO("VOLT_MULT", 32, AP_BattMonitor_AD7091R5, _volt_multiplier, 0),
+
+    // @Param: AMP_PERVLT
+    // @DisplayName: Amps per volt
+    // @Description: Number of amps that a 1V reading on the current sensor corresponds to. 
+    // @Units: A/V
+    // @User: Standard
+    AP_GROUPINFO("AMP_PERVLT", 33, AP_BattMonitor_AD7091R5, _curr_amp_per_volt, 0),
+
+    // @Param: AMP_OFFSET
+    // @DisplayName: AMP offset
+    // @Description: Voltage offset at zero current on current sensor
+    // @Units: V
+    // @User: Standard
+    AP_GROUPINFO("AMP_OFFSET", 34, AP_BattMonitor_AD7091R5, _curr_amp_offset, 0),
+
+    // @Param: VLT_OFFSET
+    // @DisplayName: Volage offset
+    // @Description: Voltage offset on voltage pin. This allows for an offset due to a diode. This voltage is subtracted before the scaling is applied
+    // @Units: V
+    // @User: Advanced
+    AP_GROUPINFO("VLT_OFFSET", 35, AP_BattMonitor_AD7091R5, _volt_offset, 0),
+    
+    // Param indexes must be less than 10 to avoid conflict with other battery monitor param tables loaded by pointer
+
+    AP_GROUPEND
+};
+
+AP_BattMonitor_AD7091R5::AnalogData AP_BattMonitor_AD7091R5::_analog_data[AD7091R5_NO_OF_CHANNELS] = {0};
+bool AP_BattMonitor_AD7091R5::_first = true;
+
+/**
+ * @brief Construct a new ap battmonitor ad7091r5::ap battmonitor ad7091r5 object
+ * 
+ * @param mon 
+ * @param mon_state 
+ * @param params 
+ */
+AP_BattMonitor_AD7091R5::AP_BattMonitor_AD7091R5(AP_BattMonitor &mon,
+                                             AP_BattMonitor::BattMonitor_State &mon_state,
+                                             AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _state.var_info = var_info;
+}
+
+/**
+ * @brief probe and initialize the sensor and register call back
+ * 
+ */
+void AP_BattMonitor_AD7091R5::init(){
+    // voltage and current pins from params and check if there are in range
+    if (_volt_pin >= 1 && _volt_pin <= AD7091R5_NO_OF_CHANNELS) {
+        volt_buff_pt = _volt_pin.get() -1;
+    }
+    if (_curr_pin >= 1 && _curr_pin <= AD7091R5_NO_OF_CHANNELS) {
+        curr_buff_pt = _curr_pin.get() -1;
+    }
+
+    if (_first) { // only the first instance read the i2c device
+        _first = false;
+        _i2c_polling = true;
+        _state.healthy = false;
+        // probe i2c device
+        _dev = hal.i2c_mgr->get_device(AD7091R5_I2C_BUS, AD7091R5_I2C_ADDR);
+        if (_dev) {
+            WITH_SEMAPHORE(_dev->get_semaphore());
+            _dev->set_retries(10); // lots of retries during probe
+            //config
+            if (_reset() && _config()) {
+                _state.healthy = true;
```

Comment:
How does the second or third instance ever get a healthy state set? I really think healthy should only be computed inside of read() if there was data read within the last 0.2 seconds.

Reply:
The thing is as all the instances are going to read data from the same I2C IC so if it is reading from one instance then it is healthy for all instances. But still, I can move to read.

Reply:
Yes, but the problem becomes the caller of the data needs to know if that specific instance is healthy, which won't be available. From a log analysis point of view you can obviously reference others, but that won't work for the internal code features.

Reply:
Don't set healthy here. Healthy implies we filled in the state. That should only be done by read(), as there won't be any valid data until after we've read it. You also don't have to set it false, because that is already handled elsewhere.